File: r_dailyprogrammer/intermediate_0104.rs
// reddit.com
// /r/dailyprogrammer/comments
// /11par4/10182012_challenge_104_intermediate_bracket_racket/

fn character_is_in_length_four_array(character: char, array: [char; 4]) -> bool {
    for &item in array.iter() {
        if item == character {
            return true
        }
    }
    false
}

fn are_matching_delimiters(first: char, second: char) -> bool {
    ((first == ('(') && second == (')')) ||
     (first == ('<') && second == ('>')) ||
     (first == ('[') && second == (']')) ||
     (first == ('{') && second == ('}')))
}

fn is_paired_correctly(expression: String) -> bool {
    let open_delimiters: [char; 4] = ['(', '<', '[', '{'];
    let close_delimiters: [char; 4] = [')', '>', ']', '}'];
    let mut unmatched: Vec<char> = Vec::new();
    for character in expression.chars() {
        if character_is_in_length_four_array(character, open_delimiters) {
            unmatched.push(character);
        }
        if character_is_in_length_four_array(character, close_delimiters) {
            let popped: Option<char> = unmatched.pop();
            let correctly_matches: bool = match popped {
                Some(opener) => are_matching_delimiters(opener, character),
                None => false
            };
            if !correctly_matches {
                return false;
            }
        }
    }
    unmatched.len() == 0
}


#[test]
fn test_is_paired_correctly() {
    assert!(is_paired_correctly(
        ".format(\"{foo}\", {'foo': \"bar\"})".to_string()));
    assert!(is_paired_correctly("([{<()>}()])".to_string()));

    assert!(!is_paired_correctly(
        "foo(''.format(\"[no pun intended]\",)".to_string()));
    assert!(!is_paired_correctly(
        "foo(''.format(\"[no pun intended]\",)))".to_string()));
    assert!(!is_paired_correctly(
        "([{<()>}()))".to_string()));
    assert!(!is_paired_correctly("(((".to_string()));
}

File: r_dailyprogrammer/intermediate_0104.py
# reddit.com
# /r/dailyprogrammer/comments
# /11par4/10182012_challenge_104_intermediate_bracket_racket/

def is_correctly_paired(expression):
    open_delimiters = ('(', '<', '[', '{')
    close_delimiters = (')', '>', ']', '}')
    close_to_match = dict(zip(close_delimiters, open_delimiters))
    unmatched = []
    for character in expression:
        if character in open_delimiters:
            unmatched.append(character)
        elif character in close_delimiters:
            if (len(unmatched) == 0 or
                close_to_match[character] != unmatched[-1]):
                return False
            else:
                unmatched.pop()
    return not unmatched

import unittest

class CorrectDelimiterPairingTestCase(unittest.TestCase):
    def test_the_light(self):
        light = (".format(\"{foo}\", {'foo': \"bar\"})",
                 "([{<()>}()])",)
        for photon in light:
            with self.subTest(i=photon):
                self.assertTrue(is_correctly_paired(photon))

    def test_the_darkness(self):
        darkness = ("foo(''.format(\"[no pun intended]\",)",
                    "foo(''.format(\"[no pun intended]\",)))",
                    "([{<()>}()))",
                    "(((",)
        for negative_luxagen in darkness:
            with self.subTest(i=negative_luxagen):
                self.assertFalse(is_correctly_paired(negative_luxagen))

if __name__ == "__main__":
    unittest.main()
